osu.Server.Spectator/Database/DatabaseAccess.cs

```diff
@@ -119,35 +119,57 @@ public async Task UpdateRoomHostAsync(MultiplayerRoom room)
             }
         }
 
-        public async Task UpdateRoomParticipantsAsync(MultiplayerRoom room)
+        public async Task AddRoomParticipantAsync(MultiplayerRoom room, MultiplayerRoomUser user)
         {
             try
             {
                 using (var transaction = await connection.BeginTransactionAsync())
                 {
-                    // This should be considered *very* temporary, and for display purposes only!
-                    await connection.ExecuteAsync("DELETE FROM multiplayer_rooms_high WHERE room_id = @RoomID", new
+                    // the user may have previously been in the room and set some scores, so need to update their presence if existing.
+                    await connection.ExecuteAsync("INSERT INTO multiplayer_rooms_high (room_id, user_id, in_room) VALUES (@RoomID, @UserID, 1) ON DUPLICATE KEY UPDATE in_room = 1", new
                     {
-                        RoomID = room.RoomID
+                        RoomID = room.RoomID,
+                        UserID = user.UserID
                     }, transaction);
 
+                    await connection.ExecuteAsync("UPDATE multiplayer_rooms SET participant_count = @Count WHERE id = @RoomID", new
+                    {
+                        RoomID = room.RoomID,
+                        Count = room.Users.Count
+                    }, transaction);
+
+                    await transaction.CommitAsync();
+                }
+            }
+            catch (MySqlException)
+            {
+                // for now we really don't care about failures in this. it's updating display information each time a user joins/quits and doesn't need to be perfect.
+            }
+        }
+
+        public async Task RemoveRoomParticipantAsync(MultiplayerRoom room, MultiplayerRoomUser user)
+        {
+            try
+            {
+                using (var transaction = await connection.BeginTransactionAsync())
+                {
                     foreach (var u in room.Users)
                     {
-                        await connection.ExecuteAsync("INSERT INTO multiplayer_rooms_high (room_id, user_id) VALUES (@RoomID, @UserID)", new
+                        await connection.ExecuteAsync("UPDATE multiplayer_rooms_high SET in_room = 0 WHERE room_id = @RoomID AND user_id = @UserID", new
                         {
                             RoomID = room.RoomID,
                             UserID = u.UserID
                         }, transaction);
                     }
```

Comment:
This can't be right, or can it? It's a loop over all users in the room, so this'll set `in_room = 0` on all of them...?

The following inspection is what tipped me off:

![2021-07-03-131219_1231x155_scrot](https://user-images.githubusercontent.com/20418176/124352501-bcc3ee80-dc00-11eb-80b8-43eef1844463.png)

Reply:
Great catch. Not having tests on the sql itself is kind of bad, which is why I wasn't merging this until actually testing live